### PR TITLE
Move form and twig to standalone bundle

### LINKED ISSUE #739: Prepare to replace by form- and twig- extensions
## Feature Request
CoreBundle was split to *-extensions and abandoned but not provide easy way to update this change in existing project. User must instant replace this bundles and all using service also upgrade all sonata dependencies at the same time but should have possibility to do it step by step.

## Summary of the feature.
CoreBundle should drop form and twig directory and required extensions 0.x created from this directory. Also extensions 0.x should include configuration from extensions 1.x. This will allow drop CoreBundle deprecations and allow use extensions 0.x instead CoreBundle. Upgrade extensions will be easy to.

## Three change are required:
- prepare form extensions 
    - [x] create form-extensions 0.x - https://github.com/sonata-project/form-extensions/pull/79
    - [ ] [release stable form-extensions 0.1](https://github.com/sonata-project/form-extensions/pull/92)
- prepare twig extensions 
    - [x] create twig-extensions 0.x - https://github.com/sonata-project/twig-extensions/pull/74
    - [ ] [release stable twig-extensions 0.1](https://github.com/sonata-project/twig-extensions/pull/80)
- [ ] replace form and twig directories by standalone extensions - #736

### PR DESCRIPTION
**Pull request by wbloszyk**
## Subject

This PR:
- drop form and twig directories
- required form- and twig- extensions (version 0.x based on dropped directories and 1.x services) 
- add deprecations notes

I am targeting this branch, because this PR is BC. 

Closes #739  

## To do
- [x] create form-extensions 0.x - https://github.com/sonata-project/form-extensions/pull/79
- [x] create twig-extensions 0.x - https://github.com/sonata-project/twig-extensions/pull/74

### CODE REVIEW

**Comment by wbloszyk** [comment]
@greg0ire Can you create 4.x branch from 3.x and target this PR to it?

**Comment by greg0ire** [comment]
Sorry I didn't see this… is this still relevant?

**Comment by wbloszyk** [comment]
My bad. I just check something. I will revert it tomorrow.

**Comment by greg0ire** [comment]
Ok. Please tell me if I still need to create a branch

**Comment by wbloszyk** [comment]
@greg0ire this PR is not required becouse: 
https://github.com/sonata-project/SonataAdminBundle/issues/6016#issuecomment-612378919

**Comment by wbloszyk** [comment]
I change concept of this PR. More information in RP description.

File: composer.json [hunk 1]
```diff
@@ -20,7 +20,9 @@
     "require": {
         "php": "^7.2",
         "cocur/slugify": "^1.4 || ^2.0 || ^3.0 || ^4.0",
-        "sonata-project/doctrine-extensions": "^1.1",
+        "sonata-project/exporter": "^1.11.0 || ^2.0",
+        "sonata-project/form-extensions": "^0.3",
+        "sonata-project/twig-extensions": "^0.2",
```
Comment (greg0ire) [inline composer.json]: Let's use 0.1-BETA for both here?

Comment (wbloszyk) [reply composer.json]: Yes, also add ^3.0:
```
"jms/serializer-bundle": "^1.0 || ^2.0 **|| ^3.0**",
```

Some test need be fix. I will do it today.

**Comment by wbloszyk** [comment]
@greg0ire 
Some tests will be fix by https://github.com/sonata-project/form-extensions/pull/89

**Comment by wbloszyk** [comment]
@core23 @phansys @greg0ire Can you check it?

BTW i use 0.x-dev extensions becouse I dont wanna w8 for release. It should be replace to tags before release.

File: composer.json [hunk 2]
```diff
@@ -67,17 +69,12 @@
     "autoload": {
         "psr-4": {
             "Sonata\\CoreBundle\\": "src/CoreBundle/",
-            "Sonata\\Form\\": "src/Form/",
-            "Sonata\\Serializer\\": "src/Serializer/",
-            "Sonata\\Twig\\": "src/Twig/"
+            "Sonata\\Serializer\\": "src/Serializer/"
```
Comment (ChristianGripp) [inline composer.json:72]: Do we need to keep this?

File: src/CoreBundle/Resources/config/twig.xml [hunk 3]
```diff
@@ -1,24 +1,30 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
     <services>
-        <service id="sonata.core.twig.extension.wrapping" class="Sonata\Twig\Extension\FormTypeExtension">
```
Comment (ChristianGripp) [inline src/CoreBundle/Resources/config/twig.xml:4]: Why did you remove this?

**Review by core23 — CHANGES_REQUESTED** [review]

File: composer.json (hunk 2 above)
Comment (wbloszyk) [reply composer.json:72]: Serializer was moved to form-extensions. We should keep sonata/serializer to keep BC.

File: src/CoreBundle/Resources/config/twig.xml (hunk 3 above)
Comment (wbloszyk) [reply src/CoreBundle/Resources/config/twig.xml:4]: I do it long time ago. I'm not sure but think twig don't allow declare the same Claas twice. Mybe we should create alias instead?

File: src/CoreBundle/Resources/config/commands.xml [hunk 4]
```diff
@@ -2,9 +2,11 @@
 <container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
     <services>
         <service id="Sonata\CoreBundle\Command\SonataDumpDoctrineMetaCommand" class="Sonata\CoreBundle\Command\SonataDumpDoctrineMetaCommand">
+            <deprecated>The "%service_id%" service is deprecated since 3.x and will be removed in 4.0.</deprecated>
```
Comment (phansys) [inline src/CoreBundle/Resources/config/commands.xml]:
```suggestion
            <deprecated>The "%service_id%" service is deprecated since sonata-project/core-bundle 3.x and will be removed in 4.0.</deprecated>
```
The same suggestion applies for the rest of deprecations.

File: tests/CoreBundle/Form/Type/BasePickerTypeTest.php [hunk 5]
```diff
@@ -74,6 +83,29 @@ public function testFinishView()
         $this->assertSame('text', $view->vars['type']);
     }
 
+    public function testTimePickerIntlFormaterWithLegacyTranslator()
```
Comment (phansys) [inline tests/CoreBundle/Form/Type/BasePickerTypeTest.php]:
```suggestion
    public function testTimePickerIntlFormaterWithLegacyTranslator(): void
```

File: tests/CoreBundle/Validator/ErrorElementTest.php [hunk 6]
```diff
@@ -229,12 +229,7 @@ public function testExceptionIsThrownWhenContextIsString()
         $constraintValidatorFactory = $this->createMock(ConstraintValidatorFactoryInterface::class);
 
         $this->expectException(
-            'InvalidArgumentException'
-        );
-        $this->expectExceptionMessage(
-            'Argument 3 passed to Sonata\Form\Validator\ErrorElement::__construct() must be an instance of '.
-            'Symfony\Component\Validator\ExecutionContextInterface or '.
-            'Symfony\Component\Validator\Context\ExecutionContextInterface.'
+            'TypeError'
```
Comment (phansys) [inline tests/CoreBundle/Validator/ErrorElementTest.php]:
```suggestion
            \TypeError::class
```

**Review by phansys — CHANGES_REQUESTED** [review]

File: src/CoreBundle/Resources/config/commands.xml (hunk 4 above)
Comment (wbloszyk) [reply src/CoreBundle/Resources/config/commands.xml]: I will do it in Monday.

File: src/CoreBundle/Resources/config/twig.xml (hunk 3 above)
Comment (ChristianGripp) [reply src/CoreBundle/Resources/config/twig.xml:4]: Keep in mind that the new bundles are not enabled by default. The classes may exist but the service do not

Comment (wbloszyk) [reply src/CoreBundle/Resources/config/twig.xml:4]: I back to home and check it in my computer. I didn't delete them. This service is under "sonata.core.twig.extension.text".

**Review by core23 — DISMISSED** [review]

File: composer.json [hunk 7]
```diff
@@ -20,7 +20,9 @@
     "require": {
         "php": "^7.2",
         "cocur/slugify": "^1.4 || ^2.0 || ^3.0 || ^4.0",
-        "sonata-project/doctrine-extensions": "^1.1",
+        "sonata-project/exporter": "^1.11.0 || ^2.0",
```
Comment (phansys) [inline composer.json]: I know the result is the same, just for the sake of consistency with other constraints:
```suggestion
        "sonata-project/exporter": "^1.11 || ^2.0",
```

File: tests/CoreBundle/Form/Type/BasePickerTypeTest.php [hunk 8]
```diff
@@ -74,6 +83,29 @@ public function testFinishView()
...
+
+        $type = new BasePickerTest(new MomentFormatConverter(), $translator);
+
+        $view = new FormView();
+        $form = new Form($this->createMock(FormConfigInterface::class));
+
+        $type->finishView($view, $form, [
+            'format' => 'H:mm',
+            'dp_min_date' => '1/1/1900',
+            'dp_max_date' => new \DateTime('3/1/2001'),
```
Comment (phansys) [inline tests/CoreBundle/Form/Type/BasePickerTypeTest.php]:
```suggestion
            'dp_min_date' => '1900-01-01',
            'dp_max_date' => new \DateTime('2001-03-01'),
```

**Review by phansys — CHANGES_REQUESTED** [review]

**Review by phansys — DISMISSED** [review]

File: composer.json [hunk 9]
```diff
@@ -20,7 +20,9 @@
     "require": {
         "php": "^7.2",
         "cocur/slugify": "^1.4 || ^2.0 || ^3.0 || ^4.0",
-        "sonata-project/doctrine-extensions": "^1.1",
+        "sonata-project/exporter": "^1.11 || ^2.0",
+        "sonata-project/form-extensions": "0.x-dev",
```
Comment (ChristianGripp) [inline composer.json:24]: RTM, but we should require stable releases

**Review by core23 — CHANGES_REQUESTED** [review]

File: composer.json (hunk 9 above)
Comment (wbloszyk) [reply composer.json:24]: https://github.com/sonata-project/SonataCoreBundle/pull/736#issuecomment-629285760

I wrote about it.

Comment (wbloszyk) [reply composer.json:24]: Done.

**Review by core23 — DISMISSED** [review]

**Comment by greg0ire** [comment]
> BTW i use 0.x-dev extensions becouse I dont wanna w8 for release. It should be replace to tags before release.

Should we use `0.1@dev` before releasing then? That way we get a green build.

**Comment by wbloszyk** [comment]
@greg0ire 
0.x-dev -work for branch, we dont have to release -BETA version. 0.x is not usable for now anyway.
0.1@dev - is for dev 0.1? or we can use `^0.1@dev` ?

File: composer.json [hunk 10]
```diff
@@ -20,7 +20,9 @@
     "require": {
         "php": "^7.2",
         "cocur/slugify": "^1.4 || ^2.0 || ^3.0 || ^4.0",
-        "sonata-project/doctrine-extensions": "^1.1",
+        "sonata-project/exporter": "^1.11 || ^2.0",
+        "sonata-project/form-extensions": "0.1",
+        "sonata-project/twig-extensions": "0.1",
```
Comment (greg0ire) [inline composer.json]: @wbloszyk please rephrase, I don't understand. I was suggesting this:

```suggestion
        "sonata-project/form-extensions": "0.1@dev",
        "sonata-project/twig-extensions": "0.1@dev",
```

because the build fails and I think it would fix it. The alternate solution would be to tag `0.1.0` for both packages, which I can do.

**Comment by greg0ire** [comment]
The build still fails… maybe the solution is `0.1-BETA@dev`, I don't know, please try locally.

**Comment by wbloszyk** [comment]
`0.1-BETA@dev` works but without last commits. We should use `0.x-dev` until make it stable or release next BETA after each PR. 

For me we should use `0.x-dev` becouse this change is in develop process and should not be used in production.

**Review by greg0ire — APPROVED** [review]

**Comment by wbloszyk** [comment]
@greg0ire I will split this PR.
First RP with deprecations and aliases should be release as fast as possibe. 
Second PR (this) will be rebase.

**Comment by greg0ire** [comment]
Let's close this one then?
